### ISSUE TITLE
Alternative image location to dockerhub

### ISSUE THREAD

**Issue by jakegt1** [kubernetes-sigs/kind#3543]
Is it possible to have the node images published somewhere else? I understand that these are under the open source licensing in dockerhub, so they aren't rate limited but for some using dockerhub is contentious (or, maybe is blocked by default due to legal wrangling.)

This is simply a selfish ask really, Kind is really simple to use for kubernetes dev but i simply cannot use it because it's in dockerhub and nowhere else. I could technically build the node image itself, but even that is a problem because the base images are.. in dockerhub. 

Understand is probably a lot of work, so if this is just a pointless ticket do just close it and leave it be. 

Thanks!

**Comment by BenTheElder** [comment]
This is a lot of work and we will probably settle on a clearer tagging scheme first, there are multiple issues open.

In the meantime, if you can access it from _anywhere_, you can mirror it yourself, and docker can support mirroring config for images from docker hub (but not other hosts).

I also work on https://registry.k8s.io, but we haven't moved kind over because of those open discussions and because this is the first time we've heard anyone say dockerhub is a problem (aside from previous rate limiting) but the other hosts can be a problem and expensive.

> I could technically build the node image itself, but even that is a problem because the base images are.. in dockerhub.

The base images themselves are not built from dockerhub except for using the golang image, if you can figure out a suitable replacement and override that you could build. It's probably much easier to just get them excepted into a mirror though.

**Comment by jakegt1** [comment]
Hmm, good to know it's only the golang image as i might be able to get that somewhere else. I have acutally built it before, so i could always write some sort of wrapper to build it and publish it myself, which will then be valid to publish. 

I can't say for other people, but i would not be surprised if other people have similar issues, but they just don't want to write about it.

**Comment by jakegt1** [comment]
hmm, i had quite a few more issues than that. I managed to build the base image by swapping the debian image, but then i ran into needing the kindnet image while building the node image itself, so guess would have to build that as well (not probably an issue.)

So is a lot of moving parts required to get it built. Of course i can workaround this and will eventually get an image, though.

**Comment by jakegt1** [comment]
Actually the latter is a hard blocker, i can build kindnet image but it expects for it to be pulled at build time (and the build itself runs in a docker container, so can't use the one i just built.)

I expect there's a way to deal with this (although probably not by using kind build node-image?), but would need more investigation on my behalf. If i've got any obvious misconceptions, would be great - because i'm happy to build it for now if i can.

**Comment by BenTheElder** [comment]
oh right, we used to sideload from the host but that doesn't work for multi-arch, so I think now you'd have to use a proxy with your mirror ... or inject mirror config

**Comment by BenTheElder** [comment]
FWIW: in docker on the host you can configure a mirror, and kind will plumb through proxy info to builds and running nodes (the env) but it will still be a bit involved ...

I think the easiest solution is to mirror the existing node images, those are single images necessary for running a cluster.

**Comment by StLeoX** [comment]
https://quay.io/repository/cilium/kindest-node?tab=tags
Maybe this is available.

**Comment by BenTheElder** [comment]
To be clear: dockerhub is the only place we're publishing currently. The above link is not provided by the kind project.

There is consideration to switch to registry.k8s.io but we won't do so until we determine the best approach (better than current) to image tagging / management because registry.k8s.io does not permit mutable tags and we want to get this right while we have the opportunity, and dockerhub is working fine for now.
